### PR TITLE
Use DeleteOptions.PropagationPolicy instead of OrphanDependents (deprecated) in kubectl

### LINKED ISSUE #59850: Use `propagationPolicy: Foreground` in kubectl by default
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
When I delete the owner object with child objects depending on it, some objects left orphaned and untouched. This is due to `kubectl` using a deprecated `orphanDependents: false` option instead of `propagationPolicy: Foreground`, and has incomplete deletion behavior in GC.

**What you expected to happen**:
All dependent objects should be either deleted immediately or marked with `deletionTimestamp` (if there is a finalizer). If children has finalizers (preventing deletion), their owner should not be deleted when `ownerReference` is marked with `blockOwnerDeletion: true` flag.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create 3 objects and set `ownerReference` with `blockOwnerDeletion: true` flag set, i.e. `root` <- `middle` <- `leaf`.
2. Add a custom finalizer (e.g. `nilebox/test-block-deletion`) on the `middle` object, which should prevent deletion of the `root`.
```
---
apiVersion: v1
kind: Secret
metadata:
  name: root
  namespace: sctest
type: Opaque
---
apiVersion: v1
kind: Secret
metadata:
  name: middle
  namespace: sctest
  finalizers: # should block deletion of parent
  - nilebox/test-block-deletion
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Secret
    name: root
    # replace UID with the actual id of "root" object
    uid: 8c9ed2c3-109c-11e8-ba18-0a517c7aa8f4
type: Opaque
---
apiVersion: v1
kind: Secret
metadata:
  name: leaf
  namespace: sctest
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Secret
    name: middle
    # replace UID with the actual id of "middle" object
    uid: 8cba4a7a-109c-11e8-ba18-0a517c7aa8f4
type: Opaque
```
3. Delete the `root` object using `kubectl` command:
```bash
delete secret root
secret "root" deleted
```
4. Result: `root` object is deleted, `middle` object is marked with `deletionTimestamp`, `leaf` is **NOT** marked with `deletionTimestamp`. 
Expected Result: `root` object is marked with `deletionTimestamp`, `middle` object is marked with `deletionTimestamp`, `leaf` is deleted.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9.1

### PR DESCRIPTION
**Pull request by nilebox**
**What this PR does / why we need it**:
Replaces the deprecated `DeleteOptions.OrphanDependents` deletion option with `DeleteOptions.PropagationPolicy`. It will improve the cascade deletion by using `Foreground` GC deletion instead of the old `OrphanDependents: false` which has a confusing behavior (leaves orphans if children has finalizers set).

**Which issue(s) this PR fixes**:
Fixes #59850

**Special notes for your reviewer**:

**Release note**:

```release-note
Use DeleteOptions.PropagationPolicy instead of OrphanDependents in kubectl 
```

### CODE REVIEW

File: pkg/kubectl/cmd/delete.go
Comment (nilebox) [reply pkg/kubectl/cmd/delete.go]: @liggitt as I noted in #59850, I would like `blockOwnerDeletion: true` flag to be respect by default deletion with `kubectl delete` command. As far as I understand, with background deletion the owner will be deleted immediately regardless of whether there are dependents with `blockOwnerDeletion: true` flag.

File: pkg/kubectl/delete.go
Comment (nilebox) [reply pkg/kubectl/delete.go]: The reasoning is the same as for generic deletion. Is there a specific reason for deleting ReplicationController immediately?

File: pkg/kubectl/cmd/delete_test.go
Comment (ash2k) [inline pkg/kubectl/cmd/delete_test.go]: No error handling on these two lines. Maybe fix it as a separate commit while at it?

Comment (ash2k) [reply pkg/kubectl/cmd/delete_test.go]: Or maybe its ok...

Comment (nilebox) [reply pkg/kubectl/cmd/delete_test.go]: I think it's fine - it's part of a unit test which will break if anything is wrong with the body.

File: pkg/kubectl/delete.go
Comment (caesarxuchao) [reply pkg/kubectl/delete.go]: At this point, the reaper has called `scaler.Scale` to stop all the dependent pods, so you can set the policy to "background" to delete the owner immediately. Moving away from `OrphanDependents` is good, as the field is marked deprecated.

Or, you can remove the reapers, and rely on the garbage collector (GC) to do the foreground deletion. The behavior will be different in that:
1. the GC deletion waits for all the dependent objects to be deleted before deleting the replicaset, not limiting to pods;
2. the GC deletion waits for the dependents to be deleted from etcd; while the scalers only wait for the pods to be in the NotReady state. GC's behavior is better, because for objects like pods, only the deletion from etcd indicates the occupied resources are released.
3.  GC cascadingly deletes dependents' dependents; reapers don't.

Moving away from reapers requires more discussion in the sig-cli, and can be done in another PR.

Comment (nilebox) [reply pkg/kubectl/delete.go]: Thanks for the context, will change it to "background" there.

File: pkg/generated/openapi/BUILD
Comment (caesarxuchao) [inline pkg/generated/openapi/BUILD]: Why is this change? Have you run hack/update-all.sh?

Comment (nilebox) [reply pkg/generated/openapi/BUILD]: no, I ran `hack/update-bazel.sh`

File: pkg/kubectl/cmd/delete.go
Comment (caesarxuchao) [inline pkg/kubectl/cmd/delete.go]: @kubernetes/sig-cli-api-reviews this is a behavior change, for a resource that doesn't have a reaper, prior to this change, it's deleted immediately by `kubectl delete --cascading=true`, now it's deleted after all its dependents are deleted. The new behavior is more similar to reaper's behavior (e.g., replicaset is deleted after it's scaled down).

File: hack/make-rules/test-cmd-util.sh
Comment (caesarxuchao) [inline hack/make-rules/test-cmd-util.sh]: Why do we need this change? The behavior of `kubectl delete deployment` at line 2861 shouldn't change.

Here and several other occurrences.

File: hack/make-rules/test-cmd-util.sh
Comment (caesarxuchao) [inline hack/make-rules/test-cmd-util.sh]: Does the podtemplate have any dependents? If not, GC should have handled the deletion quickly.

File: pkg/generated/openapi/BUILD
Comment (caesarxuchao) [reply pkg/generated/openapi/BUILD]: Could you undo the change? These diffs might be an one-time off.

File: pkg/kubectl/delete.go
Comment (caesarxuchao) [inline pkg/kubectl/delete.go]: Should be "background" here. RC is already scaled down.

Comment (caesarxuchao) [reply pkg/kubectl/delete.go]: Same for all the other reapers in this file.

File: test/e2e/apps/job.go
Comment (caesarxuchao) [inline test/e2e/apps/job.go]: This GetJob and two following checks are redundent now, please remove them.

File: test/e2e/kubectl/kubectl.go
Comment (caesarxuchao) [inline test/e2e/kubectl/kubectl.go]: ditto

**Review by caesarxuchao — COMMENTED** [review]
A few more comments. Feel free to ping me on Slack if you need help to pass the tests.

File: hack/make-rules/test-cmd-util.sh
Comment (nilebox) [reply hack/make-rules/test-cmd-util.sh]: @caesarxuchao doesn't look like it (`podtemplate.json`):
```json
  {
    "apiVersion": "v1",
    "kind": "PodTemplate",
    "metadata": {
      "name": "nginx"
    },
    "template": {
      "metadata": {
        "labels": {
          "name": "nginx"
        },
        "generateName": "nginx-"
      },
      "spec": {
         "containers": [{
           "name": "nginx",
           "image": "dockerfile/nginx",
           "ports": [{"containerPort": 80}]
         }]
      }
    }
  }
```
But the assert still fails for some reason.

**Comment by caesarxuchao** [comment]
We need further investigate the test failure at "test-cmd run_pod_templates_tests". I'll dump what I find so far.

Looking at the log, it has
```
garbagecollector.go:257] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"v1", Kind:"PodTemplate", Name:"nginx", UID:"863e5dc7-1c84-11e8-b099-0242ac110002", 
...
PodTemplate "nginx" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"foregroundDeletion"}
```

It means the garbage collector was adding the "foregroundDeletion" finalizer to nginx podTemplate, and it failed. Based on the status of the nginx podTemplate, which shows "deletingDependents:true" and "beingDeleted:true", gc should hit this function: https://github.com/kubernetes/kubernetes/blob/07240b7166d83bed49d783e0ecdfa7ee7e62cfca/pkg/controller/garbagecollector/garbagecollector.go#L392-L394, where it will remove the finalizer instead of adding one. So it doesn't explain the error in the log.

I suggest as the next step of debugging, we should run the test locally and find out where the garbage collector updates the object.

**Comment by nilebox** [comment]
> where it will remove the finalizer instead of adding one

Given that `dependents` list is empty, I agree, it should just remove the finalizer, from the log:
```
dependents:map[*garbagecollector.node]struct {}{}
```
Will try debugging.

**Comment by nilebox** [comment]
It definitely looks like a bug, but still not sure what is exactly the problem.
```
I0301 15:35:34.512031    5622 graph_builder.go:570] add [v1/PodTemplate, namespace: nislamov, name: nginx, uid: f6303b15-1d09-11e8-be6f-0800270a816f] to the attemptToDelete, because it's waiting for its dependents to be deleted
I0301 15:35:34.512079    5622 garbagecollector.go:356] processing item [v1/PodTemplate, namespace: nislamov, name: nginx, uid: f6303b15-1d09-11e8-be6f-0800270a816f]
I0301 15:35:34.514968    5622 garbagecollector.go:475] remove DeleteDependents finalizer for item [v1/PodTemplate, namespace: nislamov, name: nginx, uid: f6303b15-1d09-11e8-be6f-0800270a816f]
E0301 15:35:34.519747    5622 garbagecollector.go:257] error syncing item &garbagecollector.node{identity:garbagecollector.objectReference{OwnerReference:v1.OwnerReference{APIVersion:"v1", Kind:"PodTemplate", Name:"nginx", UID:"f6303b15-1d09-11e8-be6f-0800270a816f", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}, Namespace:"nislamov"}, dependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:1, readerWait:0}, dependents:map[*garbagecollector.node]struct {}{}, deletingDependents:true, deletingDependentsLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, beingDeleted:true, beingDeletedLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, virtual:false, virtualLock:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, owners:[]v1.OwnerReference(nil)}: PodTemplate "nginx" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"foregroundDeletion"}
```
@caesarxuchao as you correctly pointed out, according to the log GC attempted to delete the finalizer but had the request rejected because of trying to add a finalizer. Weird.
> Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"foregroundDeletion"}

File: pkg/apis/core/validation/validation.go
Comment (nilebox) [inline pkg/apis/core/validation/validation.go]: @caesarxuchao this is the bug that caused pod template test failing.
There was a wrong arguments order, see https://github.com/kubernetes/kubernetes/blob/07240b7166d83bed49d783e0ecdfa7ee7e62cfca/pkg/apis/core/validation/validation.go#L341-L343

**Comment by nilebox** [comment]
@caesarxuchao fixed the bug in API server, all tests pass now.

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: good catch. that'll need backporting

Comment (nilebox) [reply pkg/apis/core/validation/validation.go]: sure!

Comment (nilebox) [reply pkg/apis/core/validation/validation.go]: raised a PR #60683

**Comment by nilebox** [comment]
Bugfix #60683 has been merged, rebasing.

**Comment by nilebox** [comment]
@caesarxuchao @liggitt the build is green now, please take a look.

File: pkg/kubectl/delete.go
Comment (caesarxuchao) [inline pkg/kubectl/delete.go]: Could you add a comment saying we choose "background" here because the rc has already been scaled down? And for the other places in this file.

File: pkg/kubectl/cmd/delete.go
Comment (caesarxuchao) [reply pkg/kubectl/cmd/delete.go]: @liggitt @pwittrock ^^^

**Review by caesarxuchao — COMMENTED** [review]
lgtm, but will wait for the @kubernetes/sig-cli-api-reviews to approve the behavior change (https://github.com/kubernetes/kubernetes/pull/59851/files#r171069584)

File: pkg/apis/core/validation/validation.go
Comment (caesarxuchao) [reply pkg/apis/core/validation/validation.go]: Good catch! The bug prevented the garbage collector from removing finalizers because the validation code thought gc was trying to adding finalizers.

File: pkg/kubectl/cmd/delete.go
Comment (soltysh) [reply pkg/kubectl/cmd/delete.go]: +1 from me, esp. that this is the current direction we're trying to encourage.

**Comment by seans3** [comment]
@caesarxuchao and @pwittrock have asked about the behavior change, so I'm actively investigating now.

**Comment by seans3** [comment]
/lgtm
Since this change makes deletion more consistent, this behavior change appears warranted. Discussed with @pwittrock, and he agreed. A release note documenting this behavior change would be helpful.

**Comment by caesarxuchao** [comment]
My previous comment got lost since it's on the code. Repeating myself, the behavior change affects the deletion of resources that don't have a reaper. The old behavior is that the owner object is deleted immediately, and dependents are deleted in background. The new behavior is that kubectl returns immediately, but the owner object will be there, bearing a "foregroundDeletion" finalizer, until all the dependents are deleted.

**Comment by nikhiljindal** [comment]
Thanks @caesarxuchao for higlighting the change.

As per discussion in https://github.com/kubernetes/kubernetes/issues/42594, we had decided that kubectl should wait for resources to be deleted before returning. This is in violation of that.

**Comment by nilebox** [comment]
@nikhiljindal I am not sure if that's a good idea, foreground deletion could take a long time. IMO just printing a message that resource has been marked for deletion with foreground mode, and will eventually be deleted is enough.

**Comment by nilebox** [comment]
@nikhiljindal I read the linked issue and left a comment https://github.com/kubernetes/kubernetes/issues/42594#issuecomment-372840867
**TL;DR** adding a `foregroundDeletion` finalizer doesn't actually break any existing `kubectl` UX.
1. `kubectl` already doesn't wait for resources to be deleted if you add a finalizer, even for deployments and other resources with reapers.
2. For resources with reapers in this PR we use **background** deletion, so there won't be any change in behavior for these resources at all.
3. The only change in UX is the default behavior for resources that don't have reapers (`Secret`, `ConfigMap`, CRDs etc) - instead of deleting immediately we mark resource with `foregroundDeletion` finalizer, which improves the UX IMO.

With regards to #42594 I don't have any objection to `kubectl` waiting for deletion to complete (with a flag to disable this), but that's a separate issue that is already present in `master`, and this PR doesn't make things any worse. The proper implementation of #42594 will address all resources and foreground deletion shouldn't affect that.

**Comment by nikhiljindal** [comment]
> @nikhiljindal I am not sure if that's a good idea, foreground deletion could take a long time. IMO just printing a message that resource has been marked for deletion with foreground mode, and will eventually be deleted is enough.

Thats what the APIserver does. With kubectl, we want to ensure that once it returns, the resource is deleted. I agree that we have bugs and we are not there yet, but thats where we decided to want to be. https://github.com/kubernetes/kubernetes/issues/42594 has links to some background discussion.

> @nikhiljindal I read the linked issue and left a comment #42594 (comment)

Replied on the issue. I agree that we are not there yet.

> With regards to #42594 I don't have any objection to kubectl waiting for deletion to complete (with a flag to disable this), but that's a separate issue that is already present in master, and this PR doesn't make things any worse. The proper implementation of #42594 will address all resources and foreground deletion shouldn't affect that.

Agreed. I dont expect us to fix that issue completely in this PR. This PR should be good as long as it does not cause any regression.

IIUC, with foreground deletion, resources like secrets and configmaps wont be deleted immediately now. kubectl will add a finalizer which the controller will observe and remove and then delete the resource.
If yes, then thats a regression since `kubectl delete` immediately followed by `kubectl create` will fail now. Is that understanding correct?
We have kubectl tests that run create right after running delete.

**Comment by nilebox** [comment]
@nikhiljindal understood, thanks for the explanation.

That means we have to implement the wait for the foreground deletion as it was described in #42594.

**Comment by nikhiljindal** [comment]
> That means we have to implement the wait for the foreground deletion as it was described in #42594.

What if we use background instead of foreground? Will that prevent immediate deletion by adding a finalizer as well.
cc @caesarxuchao 

I agree that foreground is better than background, but we can change that in another PR after implementing wait.

**Comment by liggitt** [comment]
> What if we use background instead of foreground? Will that prevent immediate deletion by adding a finalizer as well?
> cc @caesarxuchao

GC isn't the only actor that adds finalizers. `kubectl delete` already can not (and should not) assume objects are removed from the API when the delete call returns. If we want to add wait behavior to `kubectl delete` in the future, I'd expect it to be optional and non-default.

File: pkg/kubectl/delete.go
Comment (liggitt) [inline pkg/kubectl/delete.go]: scaling down doesn't mean all dependents are deleted, just pods. why not foreground here as well?

Comment (liggitt) [reply pkg/kubectl/delete.go]: actually, can we remove all the reapers and rely on server-side GC instead at this point?

Comment (nilebox) [reply pkg/kubectl/delete.go]: > can we remove all the reapers and rely on server-side GC instead

@liggitt I think the answer is "yes", but I would prefer this change to be a separate PR.
@caesarxuchao what do you think?

Comment (nilebox) [reply pkg/kubectl/delete.go]: > why not foreground here as well?
@liggitt I would also prefer to have foreground everywhere, but @caesarxuchao previously suggested to use background for all reapers, see https://github.com/kubernetes/kubernetes/pull/59851#discussion_r171072408

**Comment by nilebox** [comment]
@soltysh PR rebased onto latest master, conflicts resolved. Do you need me to do any other changes to get this PR merged?

P.S. for some reason, `k8s-ci-bot` complains about `pull-kubernetes-unit` and refers to https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/59851/pull-kubernetes-unit/83546/, the one from 3 months ago:
```
Started | 2018-02-28 23:26 AEDT
```
Looks like `pull-kubernetes-unit` doesn't get re-executed?
Any ideas what to do about it? `/test pull-kubernetes-unit` doesn't help...

**Review by soltysh — APPROVED** [review]
/approve

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by smarterclayton** [comment]
/hold

Just looking at this, it looks like this makes `kubectl delete` for the majority of resources change from "completes and the object is deleted" to "you will now race with the GC".  That means `kubectl delete && kubectl create` no longer works correctly for anything that isn't 'pod' or 'namespace'.  Is that correct?

I'm in favor of this change, but to do it we need to make `kubectl delete` not break, which implies `wait` as discussed above.  `wait` should be the default - if we need to gather consensus to *also* make `pod` and `namespace` wait I'd be happy to arrange for that.

**Comment by nilebox** [comment]
@smarterclayton ok, I'll change default for `-wait` to be set by default in #63695

**Comment by soltysh** [comment]
@smarterclayton I agree with your approach to `--wait` but I don't want to do it here since first I want to see this code cleaned up. This includes this PR and https://github.com/kubernetes/kubernetes/pull/63979. Only then we can add nice and clean `--wait` functionality, which will happen in https://github.com/kubernetes/kubernetes/pull/63695 like Nail mentioned earlier. If we include it here, the code will be a mess, and I don't want to see that. I have a list of items that are required for 1.11 and `--wait` is already there.

**Comment by soltysh** [comment]
Let's wait for https://github.com/kubernetes/kubernetes/pull/64034 and then you'll have to rebase your PR on top of that. We'll get nice generic `--wait` mechanism which will help me with https://github.com/kubernetes/kubernetes/pull/63979.

**Comment by nilebox** [comment]
@soltysh #64034 is merged, I rebased this PR. should not be marked with `do-not-merge/hold` anymore I believe? Also need to re-add `lgtm` after rebase.

**Review by soltysh — APPROVED** [review]
/lgtm
/hold cancel

**Pull request merged**
